From 4e60aea2bb337424d3670f2076c1590d0f909cb6 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 30 Aug 2015 00:50:50 -0400
Subject: [PATCH] Pool BlockPosition to reduce memory usage

Implements a PooledMutableBlockPosition system.

Use pooled BlockPosition for many hot spots of the server that does not hold a reference past its current scope.
---
 src/main/java/net/minecraft/server/Chunk.java      |  11 +-
 src/main/java/net/minecraft/server/Entity.java     |  13 +-
 .../net/minecraft/server/NextTickListEntry.java    |   5 +
 .../net/minecraft/server/PathfinderNormal.java     |  17 +-
 .../server/PooledMutableBlockPosition.java         | 207 +++++++++++++++++++++
 src/main/java/net/minecraft/server/World.java      |  47 +++--
 .../org/bukkit/craftbukkit/block/CraftBlock.java   |  46 +++--
 7 files changed, 311 insertions(+), 35 deletions(-)
 create mode 100644 src/main/java/net/minecraft/server/PooledMutableBlockPosition.java

diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 66e1f33..4d11270 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -273,13 +273,15 @@ public class Chunk {
     }
 
     private void a(int i, int j, int k, int l) {
-        if (l > k && this.world.areChunksLoaded(new BlockPosition(i, 0, j), 16)) {
+        final PooledMutableBlockPosition pos1 = PooledMutableBlockPosition.acquire(i, 0, j); // Spigot
+        if (l > k && this.world.areChunksLoaded(pos1, 16)) { // Spigot
             for (int i1 = k; i1 < l; ++i1) {
-                this.world.c(EnumSkyBlock.SKY, new BlockPosition(i, i1, j));
+                this.world.c(EnumSkyBlock.SKY, pos1.acquireBatch(i, i1, j)); // Spigot
             }
 
             this.q = true;
         }
+        pos1.release(); // Spigot
 
     }
 
@@ -1013,6 +1015,7 @@ public class Chunk {
     }
 
     public BlockPosition h(BlockPosition blockposition) {
+        // Spigot - PooledMutableBlockPosition TODO - See if we can optimize this...
         int i = blockposition.getX() & 15;
         int j = blockposition.getZ() & 15;
         int k = i | j << 4;
@@ -1257,7 +1260,7 @@ public class Chunk {
         int k = this.g();
         boolean flag = false;
         boolean flag1 = false;
-        BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition((this.locX << 4) + i, 0, (this.locZ << 4) + j);
+        PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire((this.locX << 4) + i, 0, (this.locZ << 4) + j); // Spigot
 
         int l;
 
@@ -1272,6 +1275,7 @@ public class Chunk {
             if (!flag && i1 > 0) {
                 flag = true;
             } else if (flag && i1 == 0 && !this.world.x(blockposition_mutableblockposition)) {
+                blockposition_mutableblockposition.release(); // Spigot
                 return false;
             }
         }
@@ -1283,6 +1287,7 @@ public class Chunk {
             }
         }
 
+        blockposition_mutableblockposition.release(); // Spigot
         return true;
     }
 
diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
index c96832a..d89b2e0 100644
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -627,15 +627,15 @@ public abstract class Entity implements ICommandListener {
             int i = MathHelper.floor(this.locX);
             int j = MathHelper.floor(this.locY - 0.20000000298023224D);
             int k = MathHelper.floor(this.locZ);
-            BlockPosition blockposition = new BlockPosition(i, j, k);
+            PooledMutableBlockPosition blockposition = PooledMutableBlockPosition.acquire(i, j, k); // Spigot
             Block block = this.world.getType(blockposition).getBlock();
 
             if (block.getMaterial() == Material.AIR) {
-                Block block1 = this.world.getType(blockposition.down()).getBlock();
-
+                Block block1 = this.world.getType(blockposition.adjustY(-1)).getBlock(); // Spigot down 1 temporarily
+                blockposition.adjustY(1); // Spigot - reset back
                 if (block1 instanceof BlockFence || block1 instanceof BlockCobbleWall || block1 instanceof BlockFenceGate) {
                     block = block1;
-                    blockposition = blockposition.down();
+                    blockposition = blockposition.adjustY(-1); // Spigot
                 }
             }
 
@@ -703,6 +703,7 @@ public abstract class Entity implements ICommandListener {
                     block.a(this.world, blockposition, this); // CraftBukkit moved from above
                 }
             }
+            blockposition.release(); // Spigot
 
             // CraftBukkit start - Move to the top of the method
             /*
@@ -1450,7 +1451,7 @@ public abstract class Entity implements ICommandListener {
         if (this.noclip) {
             return false;
         } else {
-            BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition(Integer.MIN_VALUE, Integer.MIN_VALUE, Integer.MIN_VALUE);
+            PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(Integer.MIN_VALUE, Integer.MIN_VALUE, Integer.MIN_VALUE); // Spigot
 
             for (int i = 0; i < 8; ++i) {
                 int j = MathHelper.floor(this.locY + (double) (((float) ((i >> 0) % 2) - 0.5F) * 0.1F) + (double) this.getHeadHeight());
@@ -1460,10 +1461,12 @@ public abstract class Entity implements ICommandListener {
                 if (blockposition_mutableblockposition.getX() != k || blockposition_mutableblockposition.getY() != j || blockposition_mutableblockposition.getZ() != l) {
                     blockposition_mutableblockposition.c(k, j, l);
                     if (this.world.getType(blockposition_mutableblockposition).getBlock().w()) {
+                        blockposition_mutableblockposition.release(); // Spigot
                         return true;
                     }
                 }
             }
+            blockposition_mutableblockposition.release();
 
             return false;
         }
diff --git a/src/main/java/net/minecraft/server/NextTickListEntry.java b/src/main/java/net/minecraft/server/NextTickListEntry.java
index 648d255..a560929 100644
--- a/src/main/java/net/minecraft/server/NextTickListEntry.java
+++ b/src/main/java/net/minecraft/server/NextTickListEntry.java
@@ -11,6 +11,11 @@ public class NextTickListEntry implements Comparable<NextTickListEntry> {
 
     public NextTickListEntry(BlockPosition blockposition, Block block) {
         this.f = (long) (NextTickListEntry.d++);
+        // Spigot - check position - Sadly some chains can result in the position being a mutable object thats hard
+        // to prevent, so just copy when it happens.
+        if (blockposition instanceof PooledMutableBlockPosition && !((PooledMutableBlockPosition) blockposition).free) {
+            blockposition = ((PooledMutableBlockPosition) blockposition).copy();
+        }
         this.a = blockposition;
         this.e = block;
     }
diff --git a/src/main/java/net/minecraft/server/PathfinderNormal.java b/src/main/java/net/minecraft/server/PathfinderNormal.java
index 0a14c9d..4a0ead2 100644
--- a/src/main/java/net/minecraft/server/PathfinderNormal.java
+++ b/src/main/java/net/minecraft/server/PathfinderNormal.java
@@ -25,12 +25,13 @@ public class PathfinderNormal extends PathfinderAbstract {
 
         if (this.i && entity.V()) {
             i = (int) entity.getBoundingBox().b;
-            BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition(MathHelper.floor(entity.locX), i, MathHelper.floor(entity.locZ));
+            PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(MathHelper.floor(entity.locX), i, MathHelper.floor(entity.locZ)); // Spigot
 
             for (Block block = this.a.getType(blockposition_mutableblockposition).getBlock(); block == Blocks.FLOWING_WATER || block == Blocks.WATER; block = this.a.getType(blockposition_mutableblockposition).getBlock()) {
                 ++i;
                 blockposition_mutableblockposition.c(MathHelper.floor(entity.locX), i, MathHelper.floor(entity.locZ));
             }
+            blockposition_mutableblockposition.release(); // Spigot
 
             this.h = false;
         } else {
@@ -132,8 +133,8 @@ public class PathfinderNormal extends PathfinderAbstract {
 
     public static int a(IBlockAccess iblockaccess, Entity entity, int i, int j, int k, int l, int i1, int j1, boolean flag, boolean flag1, boolean flag2) {
         boolean flag3 = false;
-        BlockPosition blockposition = new BlockPosition(entity);
-        BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+        PooledMutableBlockPosition blockposition = PooledMutableBlockPosition.acquire(entity); // Spigot
+        PooledMutableBlockPosition blockposition_mutableblockposition = blockposition.acquireBatch(); // Spigot
 
         for (int k1 = i; k1 < i + l; ++k1) {
             for (int l1 = j; l1 < j + i1; ++l1) {
@@ -158,26 +159,31 @@ public class PathfinderNormal extends PathfinderAbstract {
                             flag3 = true;
                         }
 
-                        if (entity.world.getType(blockposition_mutableblockposition).getBlock() instanceof BlockMinecartTrackAbstract) {
-                            if (!(entity.world.getType(blockposition).getBlock() instanceof BlockMinecartTrackAbstract) && !(entity.world.getType(blockposition.down()).getBlock() instanceof BlockMinecartTrackAbstract)) {
+                        if (block instanceof BlockMinecartTrackAbstract) { // Spigot
+                            if (!(entity.world.getType(blockposition).getBlock() instanceof BlockMinecartTrackAbstract) && !(entity.world.getType(blockposition.down()).getBlock() instanceof BlockMinecartTrackAbstract)) { // Spigot
+                                blockposition.release(); // Spigot
                                 return -3;
                             }
                         } else if (!block.b(iblockaccess, blockposition_mutableblockposition) && (!flag1 || !(block instanceof BlockDoor) || block.getMaterial() != Material.WOOD)) {
                             if (block instanceof BlockFence || block instanceof BlockFenceGate || block instanceof BlockCobbleWall) {
+                                blockposition.release(); // Spigot
                                 return -3;
                             }
 
                             if (block == Blocks.TRAPDOOR || block == Blocks.IRON_TRAPDOOR) {
+                                blockposition.release(); // Spigot
                                 return -4;
                             }
 
                             Material material = block.getMaterial();
 
                             if (material != Material.LAVA) {
+                                blockposition.release(); // Spigot
                                 return 0;
                             }
 
                             if (!entity.ab()) {
+                                blockposition.release(); // Spigot
                                 return -2;
                             }
                         }
@@ -185,6 +191,7 @@ public class PathfinderNormal extends PathfinderAbstract {
                 }
             }
         }
+        blockposition.release(); // Spigot
 
         return flag3 ? 2 : 1;
     }
diff --git a/src/main/java/net/minecraft/server/PooledMutableBlockPosition.java b/src/main/java/net/minecraft/server/PooledMutableBlockPosition.java
new file mode 100644
index 0000000..62bc85a
--- /dev/null
+++ b/src/main/java/net/minecraft/server/PooledMutableBlockPosition.java
@@ -0,0 +1,207 @@
+/*
+ * Copyright (c) 2015. Starlis LLC / dba Empire Minecraft
+ *
+ * This source code is proprietary software and must not be redistributed without Starlis LLC's approval
+ *
+ */
+
+package net.minecraft.server;
+
+import org.apache.logging.log4j.Level;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+public final class PooledMutableBlockPosition extends BlockPosition {
+
+    private int x;
+    private int y;
+    private int z;
+
+    private int acquiredTick;
+    private Deque<PooledMutableBlockPosition> related;
+    public boolean free = false;
+    public boolean batchShifts;
+
+    private static final int MAX_SIZE = 200;
+    private static final Deque<PooledMutableBlockPosition> pool = new ArrayDeque<PooledMutableBlockPosition>(MAX_SIZE);
+
+
+    private PooledMutableBlockPosition(int x, int y, int z) {
+        super(0, 0, 0);
+        this.x = x;
+        this.y = y;
+        this.z = z;
+    }
+
+    public static PooledMutableBlockPosition acquire() {
+        return acquire(0, 0, 0);
+    }
+
+    public static PooledMutableBlockPosition acquire(Entity entity) {
+        return acquire(MathHelper.floor(entity.locX), MathHelper.floor(entity.locY), MathHelper.floor(entity.locZ));
+    }
+    public static PooledMutableBlockPosition acquire(BlockPosition pos) {
+        return acquire(pos.getX(), pos.getY(), pos.getZ());
+    }
+
+    public static synchronized PooledMutableBlockPosition acquire(int x, int y, int z) {
+        PooledMutableBlockPosition pos;
+        synchronized (pool) {
+            pos = pool.pollLast();
+        }
+        if (pos != null && pos.free) {
+            pos.free = false;
+            pos.acquiredTick = MinecraftServer.currentTick;
+            pos.set(x, y, z);
+        } else {
+            pos = new PooledMutableBlockPosition(x, y, z);
+            pos.acquiredTick = MinecraftServer.currentTick;
+        }
+
+        return pos;
+    }
+    private static void releaseToPool(PooledMutableBlockPosition pos) {
+        if (!pos.free) {
+            pos.free = true;
+            synchronized (pool) {
+                if (pool.size() < MAX_SIZE) {
+                    pool.add(pos);
+                }
+                final Deque<PooledMutableBlockPosition> rel = pos.related;
+                if (rel != null) {
+                    int allowed = MAX_SIZE - pool.size();
+                    int related = rel.size();
+                    for (int i = 0; i < allowed && i < related; i++) {
+                        final PooledMutableBlockPosition relpos = rel.poll();
+                        relpos.free = true;
+                        relpos.related = null;
+                        pool.add(relpos);
+                    }
+                    pos.related = null;
+                }
+            }
+        }
+    }
+
+    public void release() {
+        releaseToPool(this);
+    }
+
+    public PooledMutableBlockPosition acquireBatch() {
+        return acquireBatch(0, 0, 0);
+    }
+
+    public PooledMutableBlockPosition acquireBatch(EnumDirection dir) {
+        return acquireBatch(dir, 1);
+    }
+
+    public PooledMutableBlockPosition acquireBatch(EnumDirection dir, int i) {
+        return acquireBatch(x + dir.getAdjacentX() * i, y + dir.getAdjacentY() * i, z + dir.getAdjacentZ() * i);
+    }
+
+    public PooledMutableBlockPosition acquireBatch(int x, int y, int z) {
+        PooledMutableBlockPosition pos = acquire(x, y, z);
+        if (related == null) {
+            related = new ArrayDeque();
+        }
+        pos.related = related; // So any batch additions to this child goes to parent, as child won't check this
+        related.add(pos);
+        return pos;
+    }
+
+    /**
+     * We can't safely return these to pool as we don't know if its stored
+     * @param dir
+     * @param i
+     * @return
+     */
+    @Override
+    public BlockPosition shift(EnumDirection dir, int i) {
+        if (batchShifts) {
+            return acquireBatch(dir, i);
+        }
+        PooledMutableBlockPosition position = acquire(x + dir.getAdjacentX() * i, y + dir.getAdjacentY() * i, z + dir.getAdjacentZ() * i);
+        position.free = true; // Prevent returning to pool
+        return position;
+    }
+
+    public int getX() {
+        checkTick();
+        return this.x;
+    }
+
+    public int getY() {
+        checkTick();
+        return this.y;
+    }
+
+    public int getZ() {
+        checkTick();
+        return this.z;
+    }
+    private void checkTick() {
+        if (MinecraftServer.currentTick != acquiredTick) {
+            free = false;
+            new Throwable().printStackTrace();
+        }
+    }
+
+    /**
+     * This method is purely to provide compile time test incase c() doesn't get renamed
+     */
+    private static void testMutableBlockPositionMethod() {
+        BlockPosition.MutableBlockPosition test = (new BlockPosition.MutableBlockPosition()).c(0,0,0);
+    }
+    // PAIL: Rename BlockPosition.MutableBlockPosition to set - then we can remove this.
+    public PooledMutableBlockPosition c(int x, int y, int z) {
+        return set(x, y, z);
+    }
+    public PooledMutableBlockPosition set(int x, int y, int z) {
+        checkTick();
+        if (free) {
+            MinecraftServer.LOGGER.log(Level.ERROR, "PooledMutableBlockPosition modified after it was released.", new Throwable());
+            free = false;
+        }
+        this.x = x;
+        this.y = y;
+        this.z = z;
+        return this;
+    }
+
+    public PooledMutableBlockPosition set(Entity entity) {
+        return set(MathHelper.floor(entity.locX), MathHelper.floor(entity.locY), MathHelper.floor(entity.locZ));
+    }
+
+    public PooledMutableBlockPosition adjust(int x, int y, int z) {
+        return set(this.x + x, this.y + y, this.z + z);
+    }
+
+    public PooledMutableBlockPosition adjustX(int x) {
+        return set(this.x + x, this.y, this.z);
+    }
+
+    public PooledMutableBlockPosition adjustY(int y) {
+        return set(this.x, this.y + y, this.z);
+    }
+    public PooledMutableBlockPosition adjustZ(int z) {
+        return set(this.x, this.y, this.z + z);
+    }
+
+    public PooledMutableBlockPosition adjust(EnumDirection enumdirection) {
+        return adjust(enumdirection, 1);
+    }
+
+    public PooledMutableBlockPosition adjust(EnumDirection dir, int i) {
+        return set(x + dir.getAdjacentX() * i, y + dir.getAdjacentY() * i, z + dir.getAdjacentZ() * i);
+    }
+
+    @Override
+    public int compareTo(BaseBlockPosition o) {
+        return y == o.getY() ? (z == o.getZ() ? x - o.getX() : z - o.getZ()) : y - o.getY();
+    }
+
+    public BlockPosition copy() {
+        return new BlockPosition(x, y, z);
+    }
+}
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 69171a8..8a112ad 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -543,16 +543,23 @@ public abstract class World implements IBlockAccess {
 
     public void applyPhysics(BlockPosition blockposition, Block block) {
         if (captureBlockStates) { return; } // CraftBukkit - Cancel all physics during placement
+        PooledMutableBlockPosition pooled = PooledMutableBlockPosition.acquire(blockposition); // Spigot
+        pooled.batchShifts = true; // Spigot
+        blockposition = pooled; // Spigot
         this.d(blockposition.west(), block);
         this.d(blockposition.east(), block);
         this.d(blockposition.down(), block);
         this.d(blockposition.up(), block);
         this.d(blockposition.north(), block);
         this.d(blockposition.south(), block);
+        pooled.release(); // Spigot
         spigotConfig.antiXrayInstance.updateNearbyBlocks(this, blockposition); // Spigot
     }
 
     public void a(BlockPosition blockposition, Block block, EnumDirection enumdirection) {
+        PooledMutableBlockPosition pooled = PooledMutableBlockPosition.acquire(blockposition); // Spigot
+        pooled.batchShifts = true; // Spigot
+        blockposition = pooled; // Spigot
         if (enumdirection != EnumDirection.WEST) {
             this.d(blockposition.west(), block);
         }
@@ -577,6 +584,7 @@ public abstract class World implements IBlockAccess {
             this.d(blockposition.south(), block);
         }
 
+        pooled.release(); // Spigot
     }
 
     public void d(BlockPosition blockposition, final Block block) {
@@ -1180,7 +1188,7 @@ public abstract class World implements IBlockAccess {
         boolean flag = entity.aT();
         boolean flag1 = this.a(worldborder, entity);
         IBlockData iblockdata = Blocks.STONE.getBlockData();
-        BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+        //BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition(); // Spigot - not used
 
         // Spigot start
         int ystart = ( ( k - 1 ) < 0 ) ? 0 : ( k - 1 );
@@ -1283,7 +1291,7 @@ public abstract class World implements IBlockAccess {
         int l = MathHelper.floor(axisalignedbb.e + 1.0D);
         int i1 = MathHelper.floor(axisalignedbb.c);
         int j1 = MathHelper.floor(axisalignedbb.f + 1.0D);
-        BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+        PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(); // Spigot
 
         for (int k1 = i; k1 < j; ++k1) {
             for (int l1 = i1; l1 < j1; ++l1) {
@@ -1303,6 +1311,7 @@ public abstract class World implements IBlockAccess {
                 }
             }
         }
+        blockposition_mutableblockposition.release(); // Spigot
 
         return arraylist;
     }
@@ -1729,7 +1738,7 @@ public abstract class World implements IBlockAccess {
         int l = MathHelper.floor(axisalignedbb.e);
         int i1 = MathHelper.floor(axisalignedbb.c);
         int j1 = MathHelper.floor(axisalignedbb.f);
-        BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+        PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(); // Spigot
 
         for (int k1 = i; k1 <= j; ++k1) {
             for (int l1 = k; l1 <= l; ++l1) {
@@ -1737,11 +1746,13 @@ public abstract class World implements IBlockAccess {
                     Block block = this.getType(blockposition_mutableblockposition.c(k1, l1, i2)).getBlock();
 
                     if (block.getMaterial() != Material.AIR) {
+                        blockposition_mutableblockposition.release(); // Spigot
                         return true;
                     }
                 }
             }
         }
+        blockposition_mutableblockposition.release(); // Spigot
 
         return false;
     }
@@ -1753,7 +1764,7 @@ public abstract class World implements IBlockAccess {
         int l = MathHelper.floor(axisalignedbb.e);
         int i1 = MathHelper.floor(axisalignedbb.c);
         int j1 = MathHelper.floor(axisalignedbb.f);
-        BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+        PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(); // Spigot
 
         for (int k1 = i; k1 <= j; ++k1) {
             for (int l1 = k; l1 <= l; ++l1) {
@@ -1761,11 +1772,13 @@ public abstract class World implements IBlockAccess {
                     Block block = this.getType(blockposition_mutableblockposition.c(k1, l1, i2)).getBlock();
 
                     if (block.getMaterial().isLiquid()) {
+                        blockposition_mutableblockposition.release(); // Spigot
                         return true;
                     }
                 }
             }
         }
+        blockposition_mutableblockposition.release(); // Spigot
 
         return false;
     }
@@ -1779,7 +1792,7 @@ public abstract class World implements IBlockAccess {
         int j1 = MathHelper.floor(axisalignedbb.f + 1.0D);
 
         if (this.isAreaLoaded(i, k, i1, j, l, j1, true)) {
-            BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+            PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(); // Spigot
 
             for (int k1 = i; k1 < j; ++k1) {
                 for (int l1 = k; l1 < l; ++l1) {
@@ -1787,11 +1800,13 @@ public abstract class World implements IBlockAccess {
                         Block block = this.getType(blockposition_mutableblockposition.c(k1, l1, i2)).getBlock();
 
                         if (block == Blocks.FIRE || block == Blocks.FLOWING_LAVA || block == Blocks.LAVA) {
+                            blockposition_mutableblockposition.release(); // Spigot
                             return true;
                         }
                     }
                 }
             }
+            blockposition_mutableblockposition.release(); // Spigot
         }
 
         return false;
@@ -1810,7 +1825,7 @@ public abstract class World implements IBlockAccess {
         } else {
             boolean flag = false;
             Vec3D vec3d = new Vec3D(0.0D, 0.0D, 0.0D);
-            BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+            PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(); // Spigot
 
             for (int k1 = i; k1 < j; ++k1) {
                 for (int l1 = k; l1 < l; ++l1) {
@@ -1830,6 +1845,7 @@ public abstract class World implements IBlockAccess {
                     }
                 }
             }
+            blockposition_mutableblockposition.release(); // Spigot
 
             if (vec3d.b() > 0.0D && entity.aL()) {
                 vec3d = vec3d.a();
@@ -1851,17 +1867,19 @@ public abstract class World implements IBlockAccess {
         int l = MathHelper.floor(axisalignedbb.e + 1.0D);
         int i1 = MathHelper.floor(axisalignedbb.c);
         int j1 = MathHelper.floor(axisalignedbb.f + 1.0D);
-        BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+        PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(); // Spigot
 
         for (int k1 = i; k1 < j; ++k1) {
             for (int l1 = k; l1 < l; ++l1) {
                 for (int i2 = i1; i2 < j1; ++i2) {
                     if (this.getType(blockposition_mutableblockposition.c(k1, l1, i2)).getBlock().getMaterial() == material) {
+                        blockposition_mutableblockposition.release(); // Spigot
                         return true;
                     }
                 }
             }
         }
+        blockposition_mutableblockposition.release(); // Spigot
 
         return false;
     }
@@ -1873,7 +1891,7 @@ public abstract class World implements IBlockAccess {
         int l = MathHelper.floor(axisalignedbb.e + 1.0D);
         int i1 = MathHelper.floor(axisalignedbb.c);
         int j1 = MathHelper.floor(axisalignedbb.f + 1.0D);
-        BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+        PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(); // Spigot
 
         for (int k1 = i; k1 < j; ++k1) {
             for (int l1 = k; l1 < l; ++l1) {
@@ -1890,12 +1908,14 @@ public abstract class World implements IBlockAccess {
                         }
 
                         if (d0 >= axisalignedbb.b) {
+                            blockposition_mutableblockposition.release(); // Spigot
                             return true;
                         }
                     }
                 }
             }
         }
+        blockposition_mutableblockposition.release(); // Spigot
 
         return false;
     }
@@ -2383,12 +2403,13 @@ public abstract class World implements IBlockAccess {
             } else if (i >= 14) {
                 return i;
             } else {
+                PooledMutableBlockPosition mutable = PooledMutableBlockPosition.acquire(blockposition); // Spigot
                 EnumDirection[] aenumdirection = EnumDirection.values();
                 int k = aenumdirection.length;
 
                 for (int l = 0; l < k; ++l) {
                     EnumDirection enumdirection = aenumdirection[l];
-                    BlockPosition blockposition1 = blockposition.shift(enumdirection);
+                    BlockPosition blockposition1 = mutable.adjust(enumdirection); // Spigot
                     int i1 = this.b(enumskyblock, blockposition1) - j;
 
                     if (i1 > i) {
@@ -2396,9 +2417,11 @@ public abstract class World implements IBlockAccess {
                     }
 
                     if (i >= 14) {
+                        mutable.release(); // Spigot
                         return i;
                     }
                 }
+                mutable.release(); // Spigto
 
                 return i;
             }
@@ -2441,7 +2464,7 @@ public abstract class World implements IBlockAccess {
                     j2 = (l1 >> 6 & 63) - 32 + j1;
                     k2 = (l1 >> 12 & 63) - 32 + k1;
                     int l3 = l1 >> 18 & 15;
-                    BlockPosition blockposition1 = new BlockPosition(i2, j2, k2);
+                    PooledMutableBlockPosition blockposition1 = PooledMutableBlockPosition.acquire(i2, j2, k2); // Spigot
 
                     l2 = this.b(enumskyblock, blockposition1);
                     if (l2 == l3) {
@@ -2451,7 +2474,7 @@ public abstract class World implements IBlockAccess {
                             j3 = MathHelper.a(j2 - j1);
                             k3 = MathHelper.a(k2 - k1);
                             if (i3 + j3 + k3 < 17) {
-                                BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition();
+                                PooledMutableBlockPosition blockposition_mutableblockposition = PooledMutableBlockPosition.acquire(); // Spigot
                                 EnumDirection[] aenumdirection = EnumDirection.values();
                                 int i4 = aenumdirection.length;
 
@@ -2469,9 +2492,11 @@ public abstract class World implements IBlockAccess {
                                         this.H[j++] = k4 - i1 + 32 | l4 - j1 + 32 << 6 | i5 - k1 + 32 << 12 | l3 - j5 << 18;
                                     }
                                 }
+                                blockposition_mutableblockposition.release(); // Spigot
                             }
                         }
                     }
+                    blockposition1.release(); // Spigot
                 }
 
                 i = 0;
diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
index 22a8055..50e308f 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftBlock.java
@@ -24,6 +24,8 @@ import org.bukkit.metadata.MetadataValue;
 import org.bukkit.plugin.Plugin;
 import org.bukkit.util.BlockVector;
 
+import net.minecraft.server.PooledMutableBlockPosition;
+
 public class CraftBlock implements Block {
     private final CraftChunk chunk;
     private final int x;
@@ -108,7 +110,10 @@ public class CraftBlock implements Block {
     }
 
     public byte getData() {
-        IBlockData blockData = chunk.getHandle().getBlockData(new BlockPosition(x, y, z));
+        // Spigot - pooled BlockPosition
+        final PooledMutableBlockPosition position = PooledMutableBlockPosition.acquire(this.x, this.y, this.z); // Spigot
+        IBlockData blockData = chunk.getHandle().getBlockData(position);
+        position.release();
         return (byte) blockData.getBlock().toLegacyData(blockData);
     }
 
@@ -151,19 +156,35 @@ public class CraftBlock implements Block {
     @Deprecated
     @Override
     public int getTypeId() {
-        return CraftMagicNumbers.getId(chunk.getHandle().getType(new BlockPosition(this.x, this.y, this.z)));
+        // Spigot - pooled BlockPosition
+        final PooledMutableBlockPosition blockposition = PooledMutableBlockPosition.acquire(this.x, this.y, this.z);
+        final net.minecraft.server.Block type = chunk.getHandle().getType(blockposition);
+        blockposition.release();
+        return CraftMagicNumbers.getId(type);
     }
 
     public byte getLightLevel() {
-        return (byte) chunk.getHandle().getWorld().getLightLevel(new BlockPosition(this.x, this.y, this.z));
+        // Spigot - pooled BlockPosition
+        final PooledMutableBlockPosition blockposition = PooledMutableBlockPosition.acquire(this.x, this.y, this.z);
+        final byte lightLevel = (byte) chunk.getHandle().getWorld().getLightLevel(blockposition);
+        blockposition.release();
+        return lightLevel;
     }
 
     public byte getLightFromSky() {
-        return (byte) chunk.getHandle().getBrightness(EnumSkyBlock.SKY, new BlockPosition(this.x, this.y, this.z));
+        // Spigot - pooled BlockPosition
+        final PooledMutableBlockPosition blockposition = PooledMutableBlockPosition.acquire(this.x, this.y, this.z);
+        final byte brightness = (byte) chunk.getHandle().getBrightness(EnumSkyBlock.SKY, blockposition);
+        blockposition.release();
+        return brightness;
     }
 
     public byte getLightFromBlocks() {
-        return (byte) chunk.getHandle().getBrightness(EnumSkyBlock.BLOCK, new BlockPosition(this.x, this.y, this.z));
+        // Spigot - pooled BlockPosition
+        final PooledMutableBlockPosition blockposition = PooledMutableBlockPosition.acquire(this.x, this.y, this.z);
+        final byte brightness = (byte) chunk.getHandle().getBrightness(EnumSkyBlock.BLOCK, blockposition);
+        blockposition.release();
+        return brightness;
     }
 
 
@@ -359,14 +380,17 @@ public class CraftBlock implements Block {
 
     public int getBlockPower(BlockFace face) {
         int power = 0;
+        // Spigot - replace all positions with position.set()
+        final PooledMutableBlockPosition position = PooledMutableBlockPosition.acquire(this.x, this.y, this.z); // Spigot
         BlockRedstoneWire wire = Blocks.REDSTONE_WIRE;
         net.minecraft.server.World world = chunk.getHandle().getWorld();
-        if ((face == BlockFace.DOWN || face == BlockFace.SELF) && world.isBlockFacePowered(new BlockPosition(x, y - 1, z), EnumDirection.DOWN)) power = wire.getPower(world, new BlockPosition(x, y - 1, z), power);
-        if ((face == BlockFace.UP || face == BlockFace.SELF) && world.isBlockFacePowered(new BlockPosition(x, y + 1, z), EnumDirection.UP)) power = wire.getPower(world, new BlockPosition(x, y + 1, z), power);
-        if ((face == BlockFace.EAST || face == BlockFace.SELF) && world.isBlockFacePowered(new BlockPosition(x + 1, y, z), EnumDirection.EAST)) power = wire.getPower(world, new BlockPosition(x + 1, y, z), power);
-        if ((face == BlockFace.WEST || face == BlockFace.SELF) && world.isBlockFacePowered(new BlockPosition(x - 1, y, z), EnumDirection.WEST)) power = wire.getPower(world, new BlockPosition(x - 1, y, z), power);
-        if ((face == BlockFace.NORTH || face == BlockFace.SELF) && world.isBlockFacePowered(new BlockPosition(x, y, z - 1), EnumDirection.NORTH)) power = wire.getPower(world, new BlockPosition(x, y, z - 1), power);
-        if ((face == BlockFace.SOUTH || face == BlockFace.SELF) && world.isBlockFacePowered(new BlockPosition(x, y, z + 1), EnumDirection.SOUTH)) power = wire.getPower(world, new BlockPosition(x, y, z - 1), power);
+        if ((face == BlockFace.DOWN || face == BlockFace.SELF) && world.isBlockFacePowered(position.set(x, y - 1, z), EnumDirection.DOWN)) power = wire.getPower(world, position.set(x, y - 1, z), power);
+        if ((face == BlockFace.UP || face == BlockFace.SELF) && world.isBlockFacePowered(position.set(x, y + 1, z), EnumDirection.UP)) power = wire.getPower(world, position.set(x, y + 1, z), power);
+        if ((face == BlockFace.EAST || face == BlockFace.SELF) && world.isBlockFacePowered(position.set(x + 1, y, z), EnumDirection.EAST)) power = wire.getPower(world, position.set(x + 1, y, z), power);
+        if ((face == BlockFace.WEST || face == BlockFace.SELF) && world.isBlockFacePowered(position.set(x - 1, y, z), EnumDirection.WEST)) power = wire.getPower(world, position.set(x - 1, y, z), power);
+        if ((face == BlockFace.NORTH || face == BlockFace.SELF) && world.isBlockFacePowered(position.set(x, y, z - 1), EnumDirection.NORTH)) power = wire.getPower(world, position.set(x, y, z - 1), power);
+        if ((face == BlockFace.SOUTH || face == BlockFace.SELF) && world.isBlockFacePowered(position.set(x, y, z + 1), EnumDirection.SOUTH)) power = wire.getPower(world, position.set(x, y, z - 1), power);
+        position.release(); // Spigot
         return power > 0 ? power : (face == BlockFace.SELF ? isBlockIndirectlyPowered() : isBlockFaceIndirectlyPowered(face)) ? 15 : 0;
     }
 
-- 
1.9.1

